### PR TITLE
multus: allow using NADs without inspectable CIDRs

### LINKED ISSUE #12459: multus NADs do not support multiple plugins or arbitrary network types
https://github.com/rook/rook/blob/aee57ea9ffee36d680954193146d1d66c3953167/pkg/operator/ceph/config/network.go#L48-L86

This code does not allow NetworkAttachmentDefinitions to be arbitrarily specified. 

For example, the following NAD will fail to process due to the `{ "type": "sbr" }` plugin definition with an error like:
```
failed to reconcile CephCluster "rook-ceph/rook-ceph". failed to reconcile cluster "rook-ceph": failed to configure local ceph cluster: failed to create cluster: failed to start ceph monitors: failed to set Rook and/or user-defined Ceph config options after forcefully updating the existing mons: failed to generate network settings: empty subnet from network attachment definition "default/public-net"
```

A better process to follow for determining the network settings that should be used for Ceph would be to schedule a form of canary pod and observe the network annotations that get put onto the pod during runtime. Those will correctly give the CIDR that we need to use for Ceph, and it will be consistent, allowing any arbitrary NAD.

Plugin chaining is unsupported. https://www.cni.dev/plugins/current/meta/bandwidth/#chaining

It also means that the `dhcp` ipam plugin -- a reasonable option for users -- isn't supported: https://www.cni.dev/plugins/current/ipam/dhcp/

### PR DESCRIPTION
**Pull request by BlaineEXE**
**Description of your changes:**

Change how Rook detects network CIDRs for Multus networks. The IPAM
configuration is only defined as an arbitrary string JSON blob with a
"type" field and nothing more. Rook's detection of CIDRs for whereabouts
had already grown out of date since the initial implementation.
Additionally, Rook did not support DHCP IPAM, which is a reasonable
choice for users. And more, Rook did not support CNI plugin chaining,
which further complicates NADs. Based on the CNI spec, network chaning
can result in any changes to network CIDRs from the first-given plugin.

All these problems make it more and more difficult for Rook to support
Multus by inspecting the NAD itself to predict network CIDRs. Instead,
it is better for Rook to treat the CNI process as a black box. To
preserve legacy functionality of auto-detecting networks and to make
that as robust as possible, change to a canary-style architecture like
that used for Ceph mons, from which Rook will detect the network CIDRs
if possible.

Also allow users to specify overrides for CIDR ranges. This allows Rook
to still support esoteric and unexpected NAD or network configurations
where a CIDR range is not detectable or where the range detected would
be incomplete. Because it may be impossible for Rook to understand the
network CIDRs wholistically while residing only on a portion of the
network, this feature should have been present from Multus's inception.

Improving CIDR auto-detection and allowing users to specify overrides
for auto-detected CIDRs rounds out Rook's Multus support for CephCluster
(core/RADOS) installations. No further architectural changes should be
needed for CephClusters as regards application of public/cluster network
CIDRs for Multus networks.

Signed-off-by: Blaine Gardner <blaine.gardner@ibm.com>

**Which issue is resolved by this Pull Request:**
Resolves #12459 

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (BlaineEXE) [inline pkg/apis/ceph.rook.io/v1/types.go:2321]: Mostly interested in feedback about these API changes right now.

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (travisn) [inline pkg/apis/ceph.rook.io/v1/types.go:2312]: The pattern is this simple for ipv4 and v6 CIDRs? I imagined it would be more complex.

File: pkg/apis/ceph.rook.io/v1/network.go
Comment (travisn) [inline pkg/apis/ceph.rook.io/v1/network.go:75]: Would it be useful to include this `err` in the returned error below? It may help track down the parse issue.

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (travisn) [inline pkg/apis/ceph.rook.io/v1/types.go:2308]: Today users could be specifying the public/cluster CIDRs via ceph config, right? Looks good now to expose it in the cluster CR.

File: pkg/apis/ceph.rook.io/v1/network.go
Comment (BlaineEXE) [reply pkg/apis/ceph.rook.io/v1/network.go:75]: It actually doesn't end up being more useful. I looked at the output, and it's just "invalid CIDR: <address>"

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (BlaineEXE) [reply pkg/apis/ceph.rook.io/v1/types.go:2308]: Right. Currently, the multus feature knows how to figure out the CIDR from the `whereabouts` IPAM type and one or two others. But many IPAM specs don't have `range` listed, which means it can't be detected. There is also the issue that IPAM specs can list many ranges, which current code doesn't support. So this effectively acts as an overriding spec for Multus about what CIDR should be used.

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (BlaineEXE) [reply pkg/apis/ceph.rook.io/v1/types.go:2312]: Not even. The real pattern would be like 800 chars long or more. There is a comment below about the "naive" pattern

Comment (travisn) [reply pkg/apis/ceph.rook.io/v1/types.go:2312]: I see, it provides some basic validation, but leaves the more complex validation to the code helpers

File: pkg/operator/k8sutil/network.go
Comment (BlaineEXE) [inline pkg/operator/k8sutil/network.go:186]: I'm starting to consider that this is probably the wrong approach. I think we might be better of using a canary architecture like we do with the monitors to apply each NAD to a canary pod and then read the CIDRs out of the pod annotation instead. The various CNI specs linked [here](https://www.cni.dev/docs/spec/#released-versions) break down the requirements for CNI plugins and a standardized method of reporting the address ranges. I fear that continuing to go down the road of analyzing the NAD itself is merely descent into madness.

File: pkg/apis/ceph.rook.io/v1/network.go
Comment (subhamkrai) [inline pkg/apis/ceph.rook.io/v1/network.go:59]:
```suggestion
	if oldProvider != newProvider && oldProvider != "host" && newProvider != "host" {
```

I'm confused understanding this condition, could you explain here in discussion?

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (subhamkrai) [inline pkg/apis/ceph.rook.io/v1/types.go:2268]:
```suggestion
	NetworkRanges NetworkRangesSpec `json:"networkRanges,omitempty"`
```

I know this is very much self explained, but could you add one liner comment about the field

File: pkg/apis/ceph.rook.io/v1/network.go
Comment (BlaineEXE) [reply pkg/apis/ceph.rook.io/v1/network.go:59]: This is copy-pasted from some of @travisn 's code. Maybe he has a better explanation

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (BlaineEXE) [reply pkg/apis/ceph.rook.io/v1/types.go:2268]: I'm still very much not done with this code, including docs/comments cleanup, and why it's a draft right now. It's probably best to focus only on big-picture feedback right now.

File: pkg/operator/k8sutil/network.go
Comment (BlaineEXE) [reply pkg/operator/k8sutil/network.go:186]: After chatting with Multus/Whereabouts dev, I think the canary process is going to work for us. There are some pretty unlikely scenarios where CIDRs could change from node-to-node. Specifically the host-local CNI setup tends to set up different CIDRs on a per-node basis. However, because Rook expects a single NetworkAttachmentDefinition for the whole Ceph cluster, across all nodes, it makes it much less likely that someone will be able to configure a NAD with a CNI chain that give different CIDRs per-node. 

Given this known caveat, I wonder if we should consider forcing users who use host-local NADs to provide CIDRs in the CephCluster.Network spec manually. @travisn ?

We should definitely document that users with NADs that have per-node or per-zone behavior should specify the CIDRs manually in the CephCluster.Network spec.

The good thing about the canary procedure is that we allow Multus and all the CNI plugins to do their thing, and **then** get the CIDR from there, rather than trying to guess at it based on the NAD. And, the `*/network-status` annotation that Rook would use to get the CIDR(s) is strongly-typed in the CNI spec. There isn't risk that different CNIs would use different fields to get the address.

I'm going to pivot to the canary design and try to get some initial testing in, and I think I have the DNS address of a DHCP server that I can use to test (for personal use only) how DHCP reports CIDRs.

File: deploy/examples/cluster-test.yaml
Comment (BlaineEXE) [inline deploy/examples/cluster-test.yaml:33]: Note to self: TODO: remove this

**Comment by BlaineEXE** [comment]
This is ready for review, but I am adding do not merge label while I test this for DHCP networks. Results of this testing may inform some doc or minor code changes.

File: pkg/operator/ceph/cluster/mon/mon_test.go
Comment (BlaineEXE) [inline pkg/operator/ceph/cluster/mon/mon_test.go:107]: These changes may not be technically necessary, but I wasted at least 5 hours adding the operator config to this struct and all of the code places where it is called by parents before I realized that `rookVersion` was being used to hold the rook image string. I'd rather this not happen to anyone else.

File: pkg/operator/ceph/controller/network.go
Comment (BlaineEXE) [inline pkg/operator/ceph/controller/network.go:229]: I suggest reading this before reviewing.

File: pkg/operator/ceph/controller/network.go
Comment (BlaineEXE) [inline pkg/operator/ceph/controller/network.go:263]: @travisn thoughts?

File: pkg/operator/ceph/config/config.go
Comment (BlaineEXE) [inline pkg/operator/ceph/config/config.go:153]: Removed here and moved replacement to mon code because there were import chains when code existed here.

File: Documentation/CRDs/Cluster/ceph-cluster-crd.md
Comment (subhamkrai) [inline Documentation/CRDs/Cluster/ceph-cluster-crd.md:233]: For my understanding,

So users will create NAD, but they will not mention the IP's range during NAD configuration but they applying ranges laters after creation during Ceph Cluster creation?

Comment (subhamkrai) [reply Documentation/CRDs/Cluster/ceph-cluster-crd.md:233]: and users can apply multple IP' range for cluster or public network?

File: Documentation/CRDs/Cluster/ceph-cluster-crd.md
Comment (subhamkrai) [inline Documentation/CRDs/Cluster/ceph-cluster-crd.md:265]: The link is not opening for me

**Comment by subhamkrai** [comment]
Mostly, question

**Review by subhamkrai — CHANGES_REQUESTED** [review]
Some issue with crds, maybe try `make crds` again?

File: Documentation/CRDs/Cluster/ceph-cluster-crd.md
Comment (BlaineEXE) [reply Documentation/CRDs/Cluster/ceph-cluster-crd.md:233]: No, the NAD configuration hasn't changed. Users still need to configure the NAD however necessary to connect K8s pods to their network per their needs. 

Independently, users may use the `public` and `cluster` `addressRanges` to fill in the `public_network` and `cluster_network` fields of the Ceph config. Users would do this if:
1. Rook cannot auto-detect the network CIDRs from the NAD (or detects the wrong CIDRs) for some reason
2. They do not want Rook to attempt auto-detection of CIDRs from NADs
3. They are using `network.provider: host` and wish to select host interfaces for Ceph

File: Documentation/CRDs/Cluster/ceph-cluster-crd.md
Comment (BlaineEXE) [reply Documentation/CRDs/Cluster/ceph-cluster-crd.md:265]: Looks like this was a really old link that got copied. Should be fixed now.

**Comment by travisn** [comment]
The multus cluster test is failing:
```
+ sed -i 's|#deviceFilter:|deviceFilter: sdb|g' cluster-multus-test.yaml
sed: can't read cluster-multus-test.yaml: No such file or directory
Error: Process completed with exit code 2.
```

File: pkg/operator/ceph/controller/network.go
Comment (travisn) [reply pkg/operator/ceph/controller/network.go:273]: Some environments are strict about requiring resource requests/limits on all pods, so we at least need an option. The requirements are going to be very low. I'd suggest we define a new category of resources for this in the cluster CR, though by default I'd say we don't need to set them.

File: pkg/operator/ceph/controller/network.go
Comment (travisn) [inline pkg/operator/ceph/controller/network.go:351]: nit: What about a path like this? Maybe I just think of `/tmp` as a scratch folder for someone connecting to the pod.
```suggestion
		MountPath: "/var/lib/rook/multus",
```

File: deploy/examples/cluster-multus-test.yaml
Comment (travisn) [inline deploy/examples/cluster-multus-test.yaml:1]: The only multus examples are in the docs now? What about updating this example to match the doc example?

File: Documentation/CRDs/Cluster/ceph-cluster-crd.md
Comment (travisn) [inline Documentation/CRDs/Cluster/ceph-cluster-crd.md:236]: Host networking is backward compatible, we're just adding more options for public/cluster that they could also apply to host networking, right?

File: pkg/apis/ceph.rook.io/v1/network.go
Comment (travisn) [reply pkg/apis/ceph.rook.io/v1/network.go:82]: Hopefully the comment on line 79 explains it:
```
// Allow an attempt to enable or disable host networking, but not other provider changes
```

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (travisn) [inline pkg/apis/ceph.rook.io/v1/types.go:2274]: TODO

File: Documentation/CRDs/Cluster/ceph-cluster-crd.md
Comment (travisn) [inline Documentation/CRDs/Cluster/ceph-cluster-crd.md:190]: How about also updating [ceph-configuration.md](https://rook.io/docs/rook/latest/Storage-Configuration/Advanced/ceph-configuration/#osd-dedicated-network) to point to the new way of applying public/cluster settings?

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (travisn) [inline pkg/apis/ceph.rook.io/v1/types.go:2340]: Can the note be placed next to the pattern above? It seems disconnected to be several lines below.

File: pkg/operator/ceph/cluster/mon/mon_test.go
Comment (travisn) [reply pkg/operator/ceph/cluster/mon/mon_test.go:107]: Agreed, I have been confused by this as well

**Review by travisn — CHANGES_REQUESTED** [review]
Looks great overall, just some minor comments and questions.

**Comment by BlaineEXE** [comment]
Working to address feedback now, but I was able to confirm that the new auto-detection method is successfully able to get a valid CIDR when the DHCP IPAM is used. 🎉 
I am not sure how to set this up in CI. I borrowed someone's home kube lab for the day to validate that.

File: Documentation/CRDs/Cluster/ceph-cluster-crd.md
Comment (BlaineEXE) [reply Documentation/CRDs/Cluster/ceph-cluster-crd.md:190]: I had forgotten this doc section was written only as an OSD topic. I think the relevant parts of that doc are actually present in the "Ceph public and cluster networks" section that is now below, here in the CRD doc. I think it would be good to consolidate the information somewhere, and my current thought is to remove the section of the advanced config doc you linked.

Alternately, I could rewrite the Advanced config, "OSD Dedicated Network" section to contain the fuller info about public/cluster networks that I have in the CRD doc below currently. I'd then then link to the advanced doc from here, with brief info as needed. The struggle with this may be that information about `selectors` and `addressRanges` will predominantly be in a different doc, which could be confusing.

Do you have thoughts about either approach?

File: Documentation/CRDs/Cluster/ceph-cluster-crd.md
Comment (BlaineEXE) [reply Documentation/CRDs/Cluster/ceph-cluster-crd.md:236]: Right. This doc subsection is more about what public/cluster networks are, as a primer, with the `host` option detailed below.

File: deploy/examples/cluster-multus-test.yaml
Comment (BlaineEXE) [reply deploy/examples/cluster-multus-test.yaml]: This was an accidental deletion. The file **should** be back now, but let me know if you see it disappear again. I have been having trouble getting it to stay put for some reason

File: Documentation/CRDs/Cluster/ceph-cluster-crd.md
Comment (travisn) [reply Documentation/CRDs/Cluster/ceph-cluster-crd.md:190]: What if we leave that section in the advanced config untouched, except put a note at the top that says it is recommended to follow the new approach, and add a link to the new approach? I hesitate to remove the section completely since it's still valid if someone wants to use the ceph config override.

File: pkg/operator/ceph/controller/network.go
Comment (BlaineEXE) [reply pkg/operator/ceph/controller/network.go:273]: I removed this todo without resolving it in this PR. It looks like the cmd reporter doesn't set any resource requests/requirements at all, and it looks like `DetectCephVersion()` also doesn't set any resources. Can we instead take an item to handle resources for the cmd-reporter as a general work item? I think it'll bloat this more, and it seems like we don't have anyone with any big problems around this today.

**Review by subhamkrai — COMMENTED** [review]
@BlaineEXE multus ci is failing because

```
pod/rook-ceph-network-cluster-canary-mgsqm          0/1     Init:1/2   0          6m18s   10.244.246.151   fv-az1031-835   <none>           <none>
pod/rook-ceph-network-public-canary-vk4nr           0/1     Init:1/2   0          6m18s   10.244.246.150   fv-az1031-835   <none>           <none>
```
```
 status:
    conditions:
    - lastProbeTime: null
      lastTransitionTime: "2023-09-06T00:15:59Z"
      message: 'containers with incomplete status: [wait-for-network-status-annotation]'
      reason: ContainersNotInitialized
      status: "False"
      type: Initialized
````

File: Documentation/CRDs/Cluster/ceph-cluster-crd.md
Comment (BlaineEXE) [reply Documentation/CRDs/Cluster/ceph-cluster-crd.md:190]: Done, and also added a paragraph about changing network configs only with care since this section contains something similar.

File: pkg/operator/ceph/controller/network.go
Comment (travisn) [reply pkg/operator/ceph/controller/network.go]: Oh interesting, if the cmd reporter already doesn't set requests/requirements, then we can skip it for now.

File: Documentation/Storage-Configuration/Advanced/ceph-configuration.md
Comment (travisn) [inline Documentation/Storage-Configuration/Advanced/ceph-configuration.md:366]: outdated is one of the rendered keywords? I'm not even sure where the list is defined.

**Review by travisn — COMMENTED** [review]
The CI is looking good, just reran a few tests to confirm they're intermittent. Good to see the multus test passing. I'm ready to approve assuming manual testing is completed.

**Review by subhamkrai — APPROVED** [review]
I'm putting LGTM, I had a few questions that have been addressed. Thanks

File: Documentation/Storage-Configuration/Advanced/ceph-configuration.md
Comment (BlaineEXE) [reply Documentation/Storage-Configuration/Advanced/ceph-configuration.md:366]: Good point. It looks like these are the ones we can use: https://squidfunk.github.io/mkdocs-material/reference/admonitions/

I think `info`, `tip` (more catching icon than info), or `warning` could be good choices.

**Review by travisn — APPROVED** [review]

**Pull request merged**
